server/src/test/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregatorTests.java

```diff
@@ -1838,44 +1852,129 @@ public void testWithGeoPoint() throws Exception {
+                createDocument("keyword", "e", "long", 10L, "foo", "bar")
+            )
+        );
+
+        executeTestCase(true, false, new TermQuery(new Term("foo", "bar")),
+            dataset,
+            () ->
+                new CompositeAggregationBuilder("name",
+                Arrays.asList(
+                    new TermsValuesSourceBuilder("keyword").field("keyword"),
+                    new TermsValuesSourceBuilder("long").field("long")
+                )).aggregateAfter(createAfterKey("keyword", "b", "long", 10L)).size(2),
+            (result) -> {
+                assertEquals(2, result.getBuckets().size());
+                assertEquals("{keyword=c, long=100}", result.afterKey().toString());
+                assertEquals("{keyword=c, long=10}", result.getBuckets().get(0).getKeyAsString());
+                assertEquals(1L, result.getBuckets().get(0).getDocCount());
+                assertEquals("{keyword=c, long=100}", result.getBuckets().get(1).getKeyAsString());
+                assertEquals(1L, result.getBuckets().get(1).getDocCount());
+                assertTrue(result.isTerminatedEarly());
+            }
+        );
+
+        // source field and index sorting config have the different order
+        executeTestCase(true, false, new TermQuery(new Term("foo", "bar")),
+            dataset,
+            () ->
+                new CompositeAggregationBuilder("name",
+                    Arrays.asList(
+                        // reverse source order
+                        new TermsValuesSourceBuilder("keyword").field("keyword").order(SortOrder.DESC),
+                        new TermsValuesSourceBuilder("long").field("long").order(SortOrder.DESC)
+                    )
+                ).aggregateAfter(createAfterKey("keyword", "c", "long", 10L)).size(2),
+            (result) -> {
+                assertEquals(2, result.getBuckets().size());
+                assertEquals("{keyword=a, long=100}", result.afterKey().toString());
+                assertEquals("{keyword=b, long=10}", result.getBuckets().get(0).getKeyAsString());
+                assertEquals(1L, result.getBuckets().get(0).getDocCount());
+                assertEquals("{keyword=a, long=100}", result.getBuckets().get(1).getKeyAsString());
+                assertEquals(1L, result.getBuckets().get(1).getDocCount());
+                assertTrue(result.isTerminatedEarly());
+            }
+        );
+    }
+
     private void testSearchCase(List<Query> queries,
                                 List<Map<String, List<Object>>> dataset,
                                 Supplier<CompositeAggregationBuilder> create,
                                 Consumer<InternalComposite> verify) throws IOException {
         for (Query query : queries) {
-            executeTestCase(false, query, dataset, create, verify);
-            executeTestCase(true, query, dataset, create, verify);
+            executeTestCase(false, false, query, dataset, create, verify);
```

Comment:
`testEarlyTermination` uses `executeTestCase` directly and sets `useIndexSort` to true so I think we're good here. However I found a test bug when the AssertingCodec is used, this codec wraps docvalues and that breaks `DocValues#unwrapSingleton` contract. So I pushed  https://github.com/elastic/elasticsearch/pull/48399/commits/1fa763b51b69ceb3343e471a6a65981aa854f1c7 to suppress this code in the early termination tests.

Reply:
Even suppressed codecs, it seems still got the chance to use 
`keyword -> {AssertingDocValuesFormat$AssertingDocValuesProducer@4968}`. And `testEarlyTermination` would failed occasionally.

Reply:
Good catch thanks, I pushed https://github.com/elastic/elasticsearch/pull/48399/commits/b52d5c77ad718869e883de2a5fe91d9c8588fb52 to force the suppression of the Asserting codec but that means that the test is ignored when the randomization choose this codec. I need to dig a bit more to understand how we can ensure that the test always runs.

Reply:
Since system properties like` tests.codec`, `tests.docvaluesformat` and `tests.postingsformat` have been set as `random` by default in `LuceneTestCase.java`, `AssertingCode `would be selected.

It seems we could set default `Lucene80` codec directly in `executeTestCase` method as follow to solve the problem:
```
if (indexSort != null) {
      config.setCodec(TestUtil.getDefaultCodec()); // set code as default to avoid AssertingCodec
      config.setIndexSort(indexSort);
}
```